### PR TITLE
RFC #5487 remove oembed

### PR DESCRIPTION
**Pull request by Firesphere**
I only had to move the frontend embedding to ShortcodeParser.
I decided to go with it hardcoded, as the next step is remove the ShortcodeParser.

In the docs, also included the removal of RestfulService (slightly off topic, but it was the best way since I forgot to do it on the original PR where RestfulService was removed)

### CODE REVIEW

**Comment by sminnee** [comment]
This isn't quite correct.

oEmbed should still be a mandatory feature, but it should rely on the third party module.

We didn't want to drop the oEmbed feature, just to use someone else's work to power it.

File: _config.php [hunk 1]
```diff
@@ -19,7 +19,6 @@
 
 ShortcodeParser::get('default')
 	->register('file_link', array('File', 'handle_shortcode'))
-	->register('embed', array('Oembed', 'handle_shortcode'))
```
Comment (sminnee) [inline _config.php]: We'll need to have a shortcode handler that handles the bridge to the 3rd party module.

File: _config/Oembed.yml [hunk 2]
```diff
@@ -1,40 +0,0 @@
-name: Oembed
----
-Oembed:
```
Comment (sminnee) [inline _config/Oembed.yml]: Presumably the 3rd party module comes pre-loaded with this configuration? If not, we'd want to pass it in.

File: parsers/ShortcodeParser.php [hunk 3]
```diff
@@ -21,7 +23,12 @@ public function img_shortcode($attrs) {
 
 	// --------------------------------------------------------------------------------------------------------------
 
-	protected $shortcodes = array();
+	protected $shortcodes = array(
+		'embed' => array(
```
Comment (sminnee) [inline parsers/ShortcodeParser.php]: This should be provided in the config, not hard-coded into ShortcodeParser.php.

File: parsers/ShortcodeParser.php [hunk 4]
```diff
@@ -612,4 +619,51 @@ function ($matches) use ($tags, $parser) {
...
+	 *
+	 * @param $arguments
+	 * @param $content
+	 * @param $parser
+	 * @param $shortcode
+	 * @param array $extra
+	 *
+	 * @return string
+	 */
+	public static function handle_embed_shortcode($arguments, $content, $parser, $shortcode, $extra = array()) {
```
Comment (sminnee) [inline parsers/ShortcodeParser.php]: This should be in its own class, e.g. `EmbedShortcodeProvider`. Don't pollute the ShortcodeParser core with oEmbed-specific things.

File: tests/oembed/OembedTest.php [hunk 5]
```diff
@@ -1,153 +0,0 @@
-<?php
```
Comment (sminnee) [inline tests/oembed/OembedTest.php]: A test for our shortcode provider would be good.

File: docs/en/04_Changelogs/4.0.0.md [hunk 6]
```diff
@@ -278,6 +278,14 @@ E.g.
...
         'mysite/js/src/functions.js'
     ]]);
+    
+### RestfulService
+
+* `RestfulService` has been removed. Use Guzzle instead. See Upgrading notes.
+
+### Oembed
+
+* `Oembed` has been removed from the framework in favor of [oscarotero/Embed](https://github.com/oscarotero/Embed)
```
Comment (sminnee) [inline docs/en/04_Changelogs/4.0.0.md]: I'd make it clearer that oEmbed feature hasn't been dropped, just our code: "Our self-maintained Oembed implementation has been removed, in favour of introducing oscarotero/Embed as a dependency."

**Comment by sminnee** [comment]
Cool, I initially interpreted this PR incorrectly, thinking it dropped oEmbed from core. My main feedback is that the the shortcode handler belongs in its own class.

**Comment by sminnee** [comment]
> I decided to go with it hardcoded, as the next step is remove the ShortcodeParser.

Putting oEmbed-specific code will make it *harder* to swap our ShortcodeParser implementations, which is why I have suggested putting it in another class.

File: tests/oembed/OembedTest.php (hunk 5 above)
Comment (Firesphere) [inline tests/oembed/OembedTest.php]: This is the OembedTest, not the ShortcodeParser test.

File: parsers/ShortcodeParser.php (hunk 3 above)
Comment (Firesphere) [inline parsers/ShortcodeParser.php]: Same as last remark... but will fix it.

File: tests/oembed/OembedTest.php (hunk 5 above)
Comment (sminnee) [inline tests/oembed/OembedTest.php]: OK, then create `EmbedShortcodeProvider` and `EmbedShortcodeProviderTest` ;)

**Comment by Firesphere** [comment]
> Putting oEmbed-specific code will make it harder to swap our ShortcodeParser implementations, which is why I have suggested putting it in another class.

I went for this quick option, because of removing our own shortcode parser was/is literally the next on my to-do list, so I didn't really care enough about it, but agreed, even then, taking shortcuts is bad and I should feel bad about it.

File: tests/oembed/OembedTest.php (hunk 5 above)
Comment (Firesphere) [inline tests/oembed/OembedTest.php]: Jep, figured that ;)

**Comment by sminnee** [comment]
We'll still be using shortcodes in some form in the next step, so the `EmbedShortcodeProvider` will be useful there, in some form.

**Comment by Firesphere** [comment]
You want separate providers for the separate shortcodes we're going to support ourselves? Or a generic SilverStripeShortcodeProvider-like method?

**Comment by sminnee** [comment]
For now I'd just do this:

    class EmbedShortcodeProvider {
      public static handleShortcode($arguments, $content, $parser, $shortcode, $extra = array()) { ... }
      public static function embedForTemplate($embed) { ... }
    }

And in _config.php:

    ShortcodeParser::get('default')->register('embed', array('EmbedShortcodeProvider', 'handleShortcode'))

We can decide a better approach once we see what other shortcode engines expect.

**Comment by Firesphere** [comment]
Yep, cool. I think I can push that through tonight.

**Comment by Firesphere** [comment]
Updated. Implemented most feedback. 

*note:
I'll squash into single commit once it's good to merge. For now, I want to keep it on the step-by-step so the changes are better visible.

**Comment by dhensby** [comment]
Looks like something funky is going on here...

**Comment by sminnee** [comment]
Can you please rebase this branch, @Firesphere? `EmbedShortcodeProvider` and its test look pretty good. I think squashing now will be fine.

**Comment by Firesphere** [comment]
This is the rebased branch, something funky happened with rebasing, breaking a totally unrelated test it seems?

**Comment by dhensby** [comment]
Have you recessed on a stale branch or the wrong branch, there are a lot of unrelated commits here.

**Comment by Firesphere** [comment]
Re-re-re-based. This time it should be rebased correctly. I might've accidentally pulled in the wrong master branch, as there are a lot of remotes.

**Comment by Firesphere** [comment]
Error-problem discovered. My attempt to move the config to yml accidentally got pushed.

File: parsers/ShortcodeParser.php [hunk 7]
```diff
@@ -612,4 +620,10 @@ function ($matches) use ($tags, $parser) {
 
 		return $content;
 	}
+
+	public static function getShortcodes($instance = 'default')
```
Comment (sminnee) [inline parsers/ShortcodeParser.php]: Do we need this method? If not, I would remove it. I think we want to keep the number of static accessors to a minimum as they make it easy to increase coupling.

**Comment by sminnee** [comment]
This is looking really good. I just had 1 more query, commented above.

Has anyone else had a look at it?

File: parsers/ShortcodeParser.php (hunk 7 above)
Comment (Firesphere) [inline parsers/ShortcodeParser.php]: Oops.
I added that method for debugging what the hell was going on with HTMLEditorFieldTest failing. Forgot to remove it. Indeed we don't need it at all.

**Comment by dhensby** [comment]
:)

**Comment by Firesphere** [comment]
Also, Happy Birthday @dhensby

**Comment by dhensby** [comment]
Nawwwww. Thanks!

**Comment by dhensby** [comment]
Please can we remove the tests that depend on external sources. This isn't an integration test, it's a unit test. Expected input -> expected output.

if youtube or anyone else changes their responses that doesn't make our code broken it means they're breaking standards/sites.

**Pull request merged**